### DISCUSSION TITLE
Feature request/idea: Stack view

### DISCUSSION THREAD

**Discussion by remy** [maziac/DeZog#50] [Ideas]
Something I find extremely useful in cspect whilst debugging is a short stack view. I can sort of see this in dezog, but it's not straight forward. I wondered if either there was a way to either add a stack view panel or to watch words (instead of bytes).

(I thought the call stack mini panel would give me this, but it tends to be empty, so I assumed it was something else)

Call stack - I assume this isn't my code's call stack?

<img width="374" alt="Screenshot 2021-03-03 at 21 39 25" src="https://user-images.githubusercontent.com/13700/109875922-04ddff00-7c69-11eb-9d10-c624b6687bce.png">

SP in register view - tricky to see exactly what's on the stack

<img width="605" alt="Screenshot 2021-03-03 at 21 39 29" src="https://user-images.githubusercontent.com/13700/109875976-16270b80-7c69-11eb-922a-51766eb00999.png">

Cspect's version:

<img width="643" alt="Screenshot 2021-03-03 at 21 39 56" src="https://user-images.githubusercontent.com/13700/109876023-28a14500-7c69-11eb-9e94-3ba31d00b95f.png">

**Comment by maziac** [comment]
There is the LOCAL STACK in the variables section. It shows the stack of the current 'function' (call).

If that is not enough you could use a watch. It should be something Like this

sp,2

Or 

sp,2,20

**Reply by remy** [reply]
Is this the local stack (pic)
<img width="1167" alt="Screenshot 2021-03-04 at 15 06 57" src="https://user-images.githubusercontent.com/13700/109984058-533bde00-7cfb-11eb-8625-5f7b6d35d195.png">

If so - I've found it to be empty - is the a setting I need to turn on to see it (obviously after I've called).

If that does actually do the trick - then that's perfect. 

---

Ah, just added `sp 2, 5` to the watch panel - and that's exactly what I was after! 🙌 

<img width="345" alt="Screenshot 2021-03-04 at 15 09 57" src="https://user-images.githubusercontent.com/13700/109984444-b0d02a80-7cfb-11eb-881b-15fe25c3e910.png">

**Comment by maziac** [comment]
I don't know what exactly you have done / why you don't see something on the "local stack".
I made a video:
![2021-03-04 19 13 36](https://user-images.githubusercontent.com/3081973/110010405-8d6aa700-7d1e-11eb-9401-71a2aed99502.gif)


The idea is that the "local stack" show only the pushed values for the current 'function'. If you change 'function' with a call the new "local stack" is shown.
If you click on the "Call Stack" you can switch 'functions' and that switches also the "Local Stack".
(Unfortunately in that video both pushed values, HL and BC, are 0 but I think you get the point.)

**Reply by remy** [reply]
Hmm, that's definitely much more useful - but I definitely don't have that in my view 😆 . Is it possible that not setting a stack_top might be a factor? 

This is my side of things:


https://user-images.githubusercontent.com/13700/110021040-2b13a580-7d22-11eb-8653-443a0baa6a3f.mp4

**Comment by maziac** [comment]
You cannot turn the "Local Stack" on. It is always there. But you may need to expand it.

**Comment by maziac** [comment]
Strange, you should see the pushes. 
Can you maybe send me your project as a zip (or a stripped down project). And please include the sld file and the .vscode folder with launch.json.

**Reply by maziac** [reply]
Did you see my reply?

**Reply by remy** [reply]
Ah yes, sorry, got carried away.
[stack-example.zip](https://github.com/maziac/DeZog/files/6118900/stack-example.zip) - I was able to replicate with this completely pared down.

**Comment by maziac** [comment]
Tanks for reporting this.
This is actually a bug that was introduced with the 'long addresses' feature.
Normally the "topOfStack" is set to 0x10000 if it is not provided in launch.json.
Now because of 'long addresses' this was truncated to 16 bit, i.e. it was set to 0.
Which effectively meant: "no stack".

I will fix that soon.

Workaround:
It is working if you provide the "topOfStack" by yourself. E.g. try setting:
      "topOfStack": "0x5d56"
in launch.json and you should see that it is working.

It is anyhow a good idea to provide it because otherwise DeZog might misinterpret the stack and you could see weird values.
You can also set the "topOfStack" to a label so that DeZog takes the address from the SLD file.

**Reply by remy** [reply]
Ah, so that's interesting (and great that I can work around it!). Thus far I've not used `topOfStack` for anything. I'm definitely a beginner, but also I've not needed to set the `topOfStack` for any of the projects I've written (a dot command, a NextBASIC driver and a library that's loaded into a memory bank and accessed from BASIC). 

Is there some sensible default I should use - or is `0x5d56` that?

(thanks again)

**Reply by maziac** [reply]
The first thing I do at the start of a program is to set SP (after DI). Otherwise SP might point somewhere and maybe overwrite some of your code.
When you set it, you can set it to a RAM Region (defs) that you allocate.
And you can also use this as label for topOfStack.
You can have a look at the z80-sample-program for an example.

**Reply by maziac** [reply]
There is no default.
0x5d56 was just the value I found when I started to execute your program.

**Reply by maziac** [reply]
SP can be found in the sna file format. I guess sjasmplus sets it to some default by using the DEVICE directive.
(Maybe I could also automatically set topOfStack from the SNA file header, but my idea was that most probably almost all programs will any way setup there own stack just after starting, so that wouldn't help. But ... maybe I'm wrong with my guess.)

**Reply by ped7g** [reply]
SNA - since sjasmplus 1.15.0 you can specify "ramtop" value when declaring 48/128 device: `device zxspectrum48, 29999` -> which should affect stack position in SNA header (unless you produce some machine code also in ramtop-area, then emergency stack in VRAM at $4000+ is used)

NEX has SP specified in file header, no need to set it in code few instructions later, if the `.nexload` does it already (but depends how you build the NEX file, not sure what level of control NexCreator provides, sjasmplus has it as direct argument for `savenex ...`).

**Reply by remy** [reply]
I think that's the thing, thus far all my code has been for things _other than_ .sna and .nex files, so I've used ram top or top of stack :)

**Reply by maziac** [reply]
> SNA - since sjasmplus 1.15.0 you can specify "ramtop" value when declaring 48/128 device: `device zxspectrum48, 29999` -> which should affect stack position in SNA header 

@ped7g: I just tested this with 
DEVICE ZXSPECTRUM48, 0x9000

In the resulting SNA file SP is set to 0x8FFB.

(Then I found a bug in zsim. I didn't add 2 to SP before starting. Will correct that.)

Anyhow: Why is SP 0x8FFB when RAMTOP is set to 0x9000? Shouldn't it be 0x8FFD?

**Reply by maziac** [reply]
@remy So you have a few options:
- If you set the stack in the DEVICE or in the SAVENEX directive. You can use the same label also in "topOfStack" in the launch.json. (Maybe you need to subtract 2 from it, let's see what Ped's answer is)
- If you don't set anything you rely on the defaults. I.e. 0x5D58 for SNA and 0xFFFE for NEX and you can set "topOfStack" to those.
- Set the SP programmatically. In that case set the "topOfStack" to the same value you set SP to.

**Reply by maziac** [reply]
Version 2.2.3 should be available in a few moments from the marketplace. The SP has been corrected for SNA and zsim.
You can set topOfStack to 0x5D58 in launch.json if you do not set RAMTOP for the DEVICE.
